tests/integration_tests/functional/test_uffd.py

```diff
@@ -100,3 +105,70 @@ def test_unbinded_socket(bin_cloner_path, test_microvm_with_api):
     assert "Load microVM snapshot error: Cannot connect to UDS in order to" \
            " send information on handling guest memory page-faults due to: " \
            "Connection refused (os error 111)" in response.text
+
+
+def test_valid_handler(bin_cloner_path,
+                       test_microvm_with_api,
+                       uffd_handler_paths):
+    """
+    Test valid uffd handler scenario.
+
+    @type: functional
+    """
+    logger = logging.getLogger("uffd_unbinded_socket")
+
+    logger.info("Create snapshot")
+    snapshot = create_snapshot(bin_cloner_path)
+
+    logger.info("Load snapshot, mem %s", snapshot.mem)
+    vm_builder = MicrovmBuilder(bin_cloner_path)
+    vm = test_microvm_with_api
+    vm.spawn()
+
+    jailed_mem = vm.create_jailed_resource(snapshot.mem)
+
+    # Copy page fault binary into chroot of microVM.
+    jailed_handler = os.path.join(vm.chroot(), "valid_handler")
+    shutil.copyfile(uffd_handler_paths['valid_handler'], jailed_handler)
+    # Change its permissions to the UID and GID of the jail.
+    cmd = 'chown {}:{} {}'.format(
+        vm.jailer.uid, vm.jailer.gid, jailed_handler)
+    run_cmd(cmd)
+
+    real_root = os.open("/", os.O_RDONLY)
+
+    os.chroot(vm.chroot())
+    os.chdir('/')
+    st = os.stat("valid_handler")
+    os.chmod("valid_handler", st.st_mode | stat.S_IEXEC)
+    socket_path = "/firecracker-uffd.sock"
+
+    proc = Popen(
+        ["/valid_handler", socket_path, jailed_mem],
+        stdout=PIPE,
+        stderr=PIPE,
+        bufsize=1)
+    try:
+        outs, errs = proc.communicate(timeout=1)
+        print(outs)
+        print(errs)
+        assert False, "Could not start PF handler!"
+    except TimeoutExpired:
+        print("This is the good case!")
+
+    # The page fault handler will create the socket path with root rights.
+    # Change that.
+    os.chown(socket_path, vm.jailer.uid, vm.jailer.gid)
+
+    os.fchdir(real_root)
+    os.chroot(".")
+
+    vm, _ = vm_builder.build_from_snapshot(snapshot, vm=vm,
+                                           resume=True,
+                                           uffd_path=socket_path)
+
+    ssh_connection = net_tools.SSHConnection(vm.ssh_config)
+
+    # Verify if guest can run commands.
+    exit_code, _, _ = ssh_connection.execute_command("sync")
```

Comment:
should we also test for the ballooning scenario?

Reply:
To be honest, I am not sure how easy it would be to make the reproduce the balloon device scenario in our testing framework.

Reply:
wouldn't it simply be a balloon inflate -> wait -> balloon deflate?
just to see that the uffd events are handled correctly

Reply:
We chatted offline. 

It's tricky to fully control what memory regions to touch in order to trigger page faults for the pages that have been claimed by the host after a balloon inflate (and ensure they are zeroed out as expected). However, we could try to identify receiving a `UFFD_EVENT_REMOVE` after a balloon inflate. 

I will update the test.